<a href="https://colab.research.google.com/github/Dimas0824/Machine_Learning/blob/main/Jobsheet_13/ML_WEEK13.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# TUGAS PRAKTIKUM

Gunakan JST untuk klasifikasi angka tulisan tangan (MNIST).



In [16]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical

# 1. Load dataset MNIST
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 → 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# 2. Bangun model JST
model = Sequential([
    Flatten(input_shape=(28, 28)), # Ubah gambar 28x28 menjadi vektor
    Dense(128, activation='relu'), # Hidden layer 1
    Dense(64, activation='relu'),  # Hidden layer 2
    Dense(10, activation='softmax') # Output layer (10 kelas)
])

# 3. Kompilasi model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# 4. Latih model
model.fit(X_train, y_train, epochs=10, batch_size=32, validation_data=(X_test, y_test))

# 5. Evaluasi model
loss, acc = model.evaluate(X_test, y_test, verbose=0)
print(f"Akurasi pada data uji: {acc:.4f}")

/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 11s 5ms/step - accuracy: 0.8744 - loss: 0.4270 - val_accuracy: 0.9661 - val_loss: 0.1114
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step - accuracy: 0.9676 - loss: 0.1067 - val_accuracy: 0.9735 - val_loss: 0.0854
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9773 - loss: 0.0709 - val_accuracy: 0.9724 - val_loss: 0.0856
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9839 - loss: 0.0522 - val_accuracy: 0.9762 - val_loss: 0.0793
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step - accuracy: 0.9874 - loss: 0.0392 - val_accuracy: 0.9759 - val_loss: 0.0808
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step - accuracy: 0.9902 - loss: 0.0307 - val_accuracy: 0.9777 - val_loss: 0.0779
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 5ms/step - accuracy: 0.9915 - loss: 0.0260 - val_accuracy: 0.9748 - val_loss: 0.0909
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 9s 5ms/step - accuracy: 0.9932 - loss: 0.02

## Coba dengan beberapa parameter lain:

### Ubah jumlah neuron di hidden layer (misal: 256 dan 128).


In [26]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.callbacks import EarlyStopping

# 1. Load dataset MNIST
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 → 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# 2. Model JST
model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(256, activation='relu'),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

# 3. Kompilasi
model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

# 4. Early Stopping
early_stop = EarlyStopping(
    monitor='val_loss',   # berhenti jika val_loss tidak membaik
    patience=10,           # toleransi 10 epoch
    restore_best_weights=True
)

# 5. Latih model (banyak epoch)
model.fit(
    X_train, y_train,
    epochs=50,             # epoch lebih besar
    batch_size=32,
    validation_data=(X_test, y_test),
    callbacks=[early_stop]
)

# 6. Evaluasi
loss, acc = model.evaluate(X_test, y_test, verbose=0)
print(f"Akurasi pada data uji: {acc:.4f}")

Epoch 1/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 14s 7ms/step - accuracy: 0.8928 - loss: 0.3595 - val_accuracy: 0.9616 - val_loss: 0.1274
Epoch 2/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 7ms/step - accuracy: 0.9742 - loss: 0.0837 - val_accuracy: 0.9737 - val_loss: 0.0876
Epoch 3/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 7ms/step - accuracy: 0.9816 - loss: 0.0579 - val_accuracy: 0.9777 - val_loss: 0.0732
Epoch 4/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 7ms/step - accuracy: 0.9866 - loss: 0.0419 - val_accuracy: 0.9759 - val_loss: 0.0820
Epoch 5/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 7ms/step - accuracy: 0.9908 - loss: 0.0296 - val_accuracy: 0.9814 - val_loss: 0.0697
Epoch 6/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.9913 - loss: 0.0262 - val_accuracy: 0.9742 - val_loss: 0.1004
Epoch 7/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 7ms/step - accuracy: 0.9928 - loss: 0.0220 - val_accuracy: 0.9770 - val_loss: 0.0882
Epoch 8/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 7ms/step - accuracy: 0.9939 - loss: 0

### Tambahkan satu hidden layer lagi.


In [24]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.callbacks import EarlyStopping

# 1. Load dataset MNIST
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 → 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# 2. Model JST
model = Sequential([
    Flatten(input_shape=(28, 28)), # flatten image dari 28x28 menjadi vektor
    Dense(256, activation='relu'), # Hidden layer 1
    Dense(128, activation='relu'), # Hidden layer 2
    Dense(64, activation='relu'), # Hidden layer 3
    Dense(10, activation='softmax') # Output layer (10 kelas)
])

# 3. Kompilasi
model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

# 4. Early Stopping
early_stop = EarlyStopping(
    monitor='val_loss',   # berhenti jika val_loss tidak membaik
    patience=10,           # toleransi 10 epoch
    restore_best_weights=True
)

# 5. Latih model (banyak epoch)
model.fit(
    X_train, y_train,
    epochs=50,             # epoch lebih besar
    batch_size=32,
    validation_data=(X_test, y_test),
    callbacks=[early_stop]
)

# 6. Evaluasi
loss, acc = model.evaluate(X_test, y_test, verbose=0)
print(f"Akurasi pada data uji: {acc:.4f}")

Epoch 1/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 14s 7ms/step - accuracy: 0.8886 - loss: 0.3719 - val_accuracy: 0.9648 - val_loss: 0.1131
Epoch 2/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 13s 7ms/step - accuracy: 0.9735 - loss: 0.0888 - val_accuracy: 0.9748 - val_loss: 0.0834
Epoch 3/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 13s 7ms/step - accuracy: 0.9811 - loss: 0.0602 - val_accuracy: 0.9768 - val_loss: 0.0817
Epoch 4/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 13s 7ms/step - accuracy: 0.9849 - loss: 0.0471 - val_accuracy: 0.9784 - val_loss: 0.0732
Epoch 5/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 21s 7ms/step - accuracy: 0.9874 - loss: 0.0382 - val_accuracy: 0.9758 - val_loss: 0.0864
Epoch 6/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 13s 7ms/step - accuracy: 0.9887 - loss: 0.0342 - val_accuracy: 0.9741 - val_loss: 0.0950
Epoch 7/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 13s 7ms/step - accuracy: 0.9920 - loss: 0.0235 - val_accuracy: 0.9763 - val_loss: 0.0918
Epoch 8/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 14s 7ms/step - accuracy: 0.9931 - loss: 0

### Bandingkan akurasi dan waktu pelatihan.

Perbandingan Dua Model JST MNIST

| Aspek | Model 1 (256→128→Softmax) | Model 2 (256→128→64→Softmax) |
|------|---------------------------|-------------------------------|
| Jumlah Hidden Layer | 2 | 3 |
| Val Loss Terbaik | 0.0697 | 0.0732 |
| Val Accuracy Terbaik | 0.9815 | 0.9826 |
| Akurasi Akhir Test | **0.9814** | **0.9784** |
| Rata-rata Waktu per Epoch | ~12.5 detik | ~13.5 detik |
| Total Epoch (Early Stop) | 15 | 14 |
| Total Waktu Komputasi | ±187 detik (3 menit 7 detik) | ±189 detik (3 menit 9 detik) |
| Efisiensi Latihan | Lebih efisien | Sedikit lebih lambat |
| Risiko Overfitting | Lebih rendah | Lebih tinggi |



Model kedua memiliki arsitektur yang lebih dalam, sehingga jumlah parameternya lebih banyak. Untuk dataset sederhana seperti MNIST, kapasitas model yang terlalu besar menyebabkan overfitting lebih cepat, di mana model belajar pola yang tidak relevan dan kehilangan kemampuan generalisasi pada data uji. Selain itu, penambahan layer tambahan tidak memberikan informasi baru yang signifikan bagi MLP untuk gambar sederhana 28×28, sehingga justru membuat proses pelatihan lebih tidak stabil dan membutuhkan waktu komputasi lebih lama.


### Eksperimen dengan fungsi aktivasi Sigmoid vs ReLU.

Model sigmoid masih menghasilkan akurasi tinggi, namun tetap sedikit di bawah ReLU karena sigmoid lebih rentan mengalami vanishing gradient, sehingga proses belajar lebih lambat dan kadang berhenti di titik suboptimal. Secara waktu, sigmoid juga lebih lambat karena fungsi aktivasi sigmoid membutuhkan komputasi eksponensial.

 Perbandingan Model ReLU vs Sigmoid (Arsitektur Sama)

| Aspek | Model ReLU (256→128) | Model Sigmoid (256→128) |
|------|-----------------------|---------------------------|
| Aktivasi Hidden Layer | ReLU | Sigmoid |
| Epoch Total (Early Stop) | 15 | 19 |
| Val Loss Terbaik | **0.0697** | **0.0680** |
| Val Accuracy Terbaik | **0.9815** | **0.9816** |
| Akurasi Akhir Test | **0.9814** | **0.9812** |
| Rata-rata Waktu per Epoch | ~12.5 detik | ~13.5 detik |
| Total Waktu Pelatihan | ±187 detik | ±256 detik |
| Kecepatan Konvergensi | Lebih cepat | Lebih lambat |
| Risiko Vanishing Gradient | Rendah | Lebih tinggi |
| Kompleksitas Komputasi | Ringan | Lebih berat (fungsi sigmoid) |



In [27]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.callbacks import EarlyStopping

# 1. Load dataset MNIST
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 → 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# 2. Model JST
model = Sequential([
    Flatten(input_shape=(28, 28)), # flatten image dari 28x28 menjadi vektor
    Dense(256, activation='sigmoid'), # Hidden layer 1
    Dense(128, activation='sigmoid'), # Hidden layer 2
    Dense(10, activation='softmax') # Output layer (10 kelas)
])

# 3. Kompilasi
model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

# 4. Early Stopping
early_stop = EarlyStopping(
    monitor='val_loss',   # berhenti jika val_loss tidak membaik
    patience=10,           # toleransi 10 epoch
    restore_best_weights=True
)

# 5. Latih model (banyak epoch)
model.fit(
    X_train, y_train,
    epochs=50,             # epoch lebih besar
    batch_size=32,
    validation_data=(X_test, y_test),
    callbacks=[early_stop]
)

# 6. Evaluasi
loss, acc = model.evaluate(X_test, y_test, verbose=0)
print(f"Akurasi pada data uji: {acc:.4f}")

Epoch 1/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 14s 7ms/step - accuracy: 0.8054 - loss: 0.7042 - val_accuracy: 0.9432 - val_loss: 0.1885
Epoch 2/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 13s 7ms/step - accuracy: 0.9494 - loss: 0.1700 - val_accuracy: 0.9594 - val_loss: 0.1329
Epoch 3/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.9672 - loss: 0.1102 - val_accuracy: 0.9665 - val_loss: 0.1012
Epoch 4/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.9773 - loss: 0.0750 - val_accuracy: 0.9736 - val_loss: 0.0830
Epoch 5/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.9837 - loss: 0.0575 - val_accuracy: 0.9765 - val_loss: 0.0733
Epoch 6/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 13s 7ms/step - accuracy: 0.9876 - loss: 0.0414 - val_accuracy: 0.9769 - val_loss: 0.0750
Epoch 7/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 6ms/step - accuracy: 0.9903 - loss: 0.0314 - val_accuracy: 0.9769 - val_loss: 0.0756
Epoch 8/50
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 12s 7ms/step - accuracy: 0.9937 - loss: 0